picolibc: tests with meson

domin144 opened the pull request:
Switch from manually building picolibc tests to bilding and running them
as part of picolibc build. This enables tests, which were not included
in the manual build.

Several fixes were necessary:

1. Add '-nostdlib' to picolibc c_args
This prevents errors like:
  ld.lld: error: unable to find library -lc
in checks for compiler flags support in picolibs build scripts.
Example of failing check:
  has_link_defsym = meson.get_cross_property('has_link_defsym', cc.has_link_argument('-Wl,--defsym=' + global_prefix + '_start=0'))

2. Make picolibc use compiler_rt in place of libgcc.

3. Change machines used with qemu and adjust memory layouts.
Some of picolibc tests need more memory, than previous machines
provided. Example error:
  ld.lld: error: section '.text' will not fit in region 'flash': overflowed by 34504 bytes

4. Postpone building picolibc tests to the test step. compiler_rt
requires picolibc to be installed. picolibc tests require compiler_rt.
The delayed tests enable picolibc to be installed before compiler_rt is built.

mplatings reviewed (commented):
Looks good overall, thanks

domin144 commented on CMakeLists.txt:
```diff
@@ -946,88 +911,101 @@ add_library_variant(
     COMPILE_FLAGS "-march=armv8-a"
     MULTILIB_FLAGS "--target=aarch64-none-unknown-elf"
     QEMU_PARAMS "-M virt -cpu cortex-a57"
+    MEMORY_LAYOUT 0x40000000 4M 0x40400000 2M 8k
```
Done

domin144 commented on cmake/to_meson_list.cmake:
```diff
@@ -0,0 +1,11 @@
+    list(TRANSFORM input_list APPEND "'")
+    list(JOIN input_list ", " input_list)
+    set(${target_variable} "${input_list}" PARENT_SCOPE)
+endfunction()
```
Done

domin144 commented on test-support/run-picolibc-test.py:
```diff
@@ -0,0 +1,88 @@
+import subprocess
+
+# This script requires Python 3.6 or later
+assert sys.version_info >= (3, 6)
```
Done

domin144 commented:
I applied corrections for the comments.
I also removed the linker scripts and replaced with default ones, which now have values matching our test platforms.

mplatings commented on test-support/lit-exec-qemu.py:
```diff
@@ -26,7 +26,11 @@ def run(args):
     qemu_cmd = ([args.qemu_command] + qemu_params +
-        ['-semihosting', '-nographic', '-kernel', image])
+        ['-semihosting', '-nographic'])
+    if qemu_params.count("virt"):
```
Rather than trying to parse the qemu parameters, pull the machine type out into a separate argument that is parsed with argparse.

mplatings commented on test-support/run-picolibc-test.py:
```diff
@@ -0,0 +1,85 @@
+            "-monitor", "none",
+            "-serial", "none",
+            "-nographic"]
+    if qemu_params.count("virt"):
```
Rather than trying to parse the qemu parameters, pull the machine type out into a separate argument that is parsed with argparse.

domin144 replied:
Done in https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/pull/256

mplatings commented on test-support/run-picolibc-test.py:
```diff
@@ -0,0 +1,85 @@
+    if image_args[0:1] == ['--']:
+        image_args = image_args[1:]
+    if is_disabled(image):
+        return 77
```
Presumably this is a magic signal to Meson. Please add a comment to make that explicitly clear.

domin144 replied:
Done

mplatings commented on test-support/run-picolibc-test.py:
```diff
@@ -0,0 +1,85 @@
+    assert command
+    image = command[0]
+    image_args = command[1:]
+    if image_args[0:1] == ['--']:
```
if image_args[0] == "--":

domin144 replied:
I found out, that argparse already handles "--" the way we need it. Removed.

mplatings commented on test-support/run-picolibc-test.py:
```diff
@@ -0,0 +1,85 @@
+def is_disabled(image):
+    return any([image.endswith(t) for t in disabled_tests])
+
+def run(args):
```
This function contains complex parsing logic. Please pull that out into a separate function with a doctest so that anyone modifying it later will be able to check they haven't broken it. Ideally use argparse to do the parsing.

domin144 replied:
I simplified the function. Argparse does all parsing now.

mplatings commented on CMakeLists.txt:
```diff
@@ -460,16 +472,32 @@ function(add_picolibc directory variant target_triple flags qemu_params variant_
+            -Dlibdir=lib
+            -Dspecsdir=none
+            -Dmultilib=false
+            -Dtests-enable-stack-protector=false
```
Thanks for explaining. Let's keep it then.

domin144 commented:
I addressed some of the issues in a separate PR https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/pull/256 to make changes reviewable.
I will rebase this PR, when those changes are merged.

domin144 commented on CMakeLists.txt:
```diff
@@ -864,12 +817,28 @@ function(get_compiler_rt_target_triple target_arch flags)
+    set(flash_size ${VARIANT_FLASH_SIZE})
+    set(ram_address ${VARIANT_RAM_ADDRESS})
+    set(ram_size ${VARIANT_RAM_SIZE})
+    set(stack_size ${VARIANT_STACK_SIZE})
```
Done in https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/pull/256

domin144 commented on CMakeLists.txt:
```diff
@@ -705,15 +635,38 @@ Hard-float library with target triple \"${target_triple}\" must end \"-eabihf\""
+        -DLIBC_LINKER_SCRIPT=picolibcpp.ld
+        -DLIBCXX_EXECUTOR=${test_executor}
+        -DLIBCXXABI_EXECUTOR=${test_executor}
+        -DLIBUNWIND_EXECUTOR=${test_executor}
```
Done in https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/pull/256

mplatings commented on test-support/lit-exec-qemu.py:
```diff
@@ -5,15 +5,53 @@
+# https://mesonbuild.com/Unit-tests.html#skipped-tests-and-hard-errors
+EXIT_CODE_SKIP = 77
+
+disabled_tests = [
```
This information doesn't belong in this file.

mplatings commented on test-support/lit-exec-qemu.py:
```diff
@@ -5,15 +5,53 @@
 # This script is a bridge between lit-based tests of LLVM C++ runtime libraries
 # (libc++abi, libunwind, libc++) and QEMU. It must handle the same command-line
 # arguments as llvm-project/libcxx/utils/run.py.
+# This is also a wrapper script to run picolibc tests with QEMU.
```
This goes against the philosophy of "Make each program do one thing well."

mplatings commented on CMakeLists.txt:
```diff
@@ -488,134 +509,37 @@ function(
+    # the full target, which would include tests. Picolibc tests, are enabled
+    # only in tests step, otherwise, they would be built before install.
+    add_custom_target(compiler_rt_${variant}-available)
+    ExternalProject_Add_StepDependencies(picolibc_${variant} test compiler_rt_${variant}-available)
```
Could a standard step be used instead? e.g. `compiler_rt_${variant}-install`

domin144 replied:
It would be more readable, but the issue is, the target has to exist at the point where `ExternalProject_Add_StepDependencies` is called and `compiler_rt_${variant}-install` is created later, by `ExternalProject` call for `compiler_rt`. I added this dummy `compiler_rt_${variant}-available` to resolve the mutual dependency between `picolibc` and `compiler-rt`.

mplatings replied:
> the target has to exist at the point where ExternalProject_Add_StepDependencies is called

Are you sure? Usually in CMake targets don't have to exist at the point the dependency is declared.

domin144 replied:
You are right. CMake happily accepted the `-install` target. I will fix this in next iteration.

mplatings reviewed (changes requested):
Almost there!

domin144 commented:
I uploaded a new patch. I won't be able to address new comments till Tuesday.

mplatings reviewed (approved):
Looks good, thanks!
